accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+
+Rationale: Allow components update as soon as it is needed for the software component, without waiting for the general maintenance update.
+
+**Isolation**
+
+Since Salt Minion is a part of configuration management system, it is also a good to have it "static", in terms of zero dependencies from the operating system it currently runs.
+
+Rationale: Allow running latest modern software on much older environments without any impact to the managed operating system.
+
+**In-place Bootstrapping**
+
+Each Salt Minion is not packaged anymore in OS-specific package manager (RPM, Deb etc), but is bootstrapped via standard Python mechanism (Package Installer for Python, PIP).
+
+Rationale: Track all the patches in only one place. As the code is isolated (see above), avoid unnecessary extra-packaging for isolated version of the Salt Minion.
+
+## Principle
+
+The principle is to run own minimal [PyPi](https://pypi.org/project/pypiserver/) local server on Uyuni cluster. In order to install a Salt Minion, a managed system should use PIP againstlocal PyPi server, as shown on Figure 1 below:
+
+![Layout](images/00065-layout.png)
+
+_Figure 1_
+
+Process update or installation flow as follows:
+1. One RPM general package of Salt Minion **(1)** is provided for all possible platforms and architectures and is installed on PyPi server. However, Salt Minion is not _installed_ per se in the sense as for end-user, but as a PyPi package into the PyPi repository location. This way patches are still tracked at RPM `.spec` level.
+2. General statically compiled Python vanilla environment is provided as an RPM package from Uyuni repository and is allowed to be installed **(2)** on a Managed System, e.g. in `/opt/python3.8` or whatever version is needed. That way it is also allowed to have many Python versions as well as its modifications, if ever needed.
+3. Once virtual environment (not shown on the _Figure 1_) is created for the the separate Python interpreter, PIP is bootstrapping the entire Salt Minion against Uyuni-served PyPi server with all the required Salt Minion dependencies.
+
+This should result into running fully featured Salt Minion of latest version in an isolated virtual environment on an OS-isolated Python interpreter, which is aligned with the current Uyuni Server. Once Uyuni Server is updated to another version, so is the PyPi container image is also upgraded. Thus, Uyuni Server should issue mass-upgrade to all Salt Minions that should PIP-upgrade themselves, bumping up to the latest available version.
+
+Consequently, this mechanism will allow to keep always aligned Salt Minions across the entire infrastructure consistent and identical.
+
+## Fail-proof Updates
+
+Having running Salt Minion inside a virtual environment, allows to deliver fail-proof updates. The basic mechanism is to clone Python Virtual Environment from the current active Salt Minion and update the Salt Minion there.
```

Comment:
Actually I was wondering about this.

With venvjail we created a tgz for each architecture. All of them lives in the master node and are scp to the target. From there are untarged and the systemd service was restarted (pointing into the new venv).

The goal was exactly this: rollback. But you need to implement all the mechanism of venv distribution, fail detection and rollback. We used ansible, so not agent was required, but I wonder how this can be implemented in salt, when we are updating the same agent that will do the replacement, the fail detection and the rollback.

Reply:
Use Ansible to install Salt? 😄 Well, as I mentioned earlier, a venv manager would be needed to implement. Something that would a) prepare static Python and b) create/manage venvs and relink them for systemd. Actually, Ansible won't "fly" either, if there is no correct Python yet at all installed.

Reply:
: D I do not know!

Can you attach into the RPF this workflow? And from where the metadata required to build the venv comes from? For example, how this agent living in the node (the venv manager) will discover all the data that requires to bootstrap the first venv.